include/dsn/utility/ringbuf.h

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2017, Xiaomi, Inc.  All rights reserved.
+// This source code is licensed under the Apache License Version 2.0, which
+// can be found in the LICENSE file in the root directory of this source tree.
+
+#pragma once
+
+/*
+Example：
+
+int main()
+{
+    ringbuf<int, 5> q;
+    for (int i=1;i<=7;i++){
+        q.push(i);
+        std::cout<<q.front()<<" "<<q.back()<<" "<<q.size()<<std::endl;
+    }
+    std::cout<<"-------"<<std::endl;
+    while (!q.empty()){
+        std::cout<<q.front()<<" "<<q.back()<<" "<<q.size()<<std::endl;
+        q.pop();
+    }
+}
+
+Result:
+1 1 1
+1 2 2
+1 3 3
+1 4 4
+1 5 5
+2 6 5
+3 7 5
+-------
+3 7 5
+4 7 4
+5 7 3
+6 7 2
+7 7 1
+
+*/
+
+namespace dsn {
+namespace utils {
+
+template <typename Type, int MaxElements>
+class ringbuf
+{
```

Comment:
This class is not thread-safe, it's necessary to add some comment.
And you want to use it for read/write operation hotkey detection, right? Is there any problem if you use it in multi-threads scenario?

Reply:
It's just like `std::queue`, we should add a mutex to keep safety in a multi-threads scenario.